### PR TITLE
For FullCBFs, try underload before falling back on undefined

### LINKED ISSUE #559: FullCBF trusted range
When reviewing #536 I missed that it would affect XFEL CI test [tst_cxi_formatXTC.py](https://gitlab.com/cctbx/xfel_regression/-/blob/main/test/command_line/tst_cxi_formatXTC.py) (now disabled), but as I was looking into fixing it it sent me waaaaay down a rabbit hole.  Our current implementations of Full CBFs use conventions from older versions of ImageCIF, but newer versions have clarified overload and underload in a way that we may want to account for.  Specifically, these are differences between between version 1.5.4 and 1.8.4 (when exactly the changes were added I didn't dig into).

1. ImageCIF 1.8.4 specifies [`underload`](http://www.bernstein-plus-sons.com/software/CBF/doc/cif_img_1.8.4.html#_array_intensities.underload) as "The lowest value at which pixels for this detector are measured."  However our full CBF readers use [`undefined_value`](http://www.bernstein-plus-sons.com/software/CBF/doc/cif_img_1.8.4.html#_array_intensities.undefined_value), defined as "A value to be substituted for undefined values in the data array."  The dxtbx code is [here](https://github.com/cctbx/dxtbx/blob/main/src/dxtbx/model/detector_helpers.py#L127-L132), used [here](https://github.com/cctbx/dxtbx/blob/main/src/dxtbx/model/detector.py#L777).  Note that `underload` is not in [1.5.4](http://www.bernstein-plus-sons.com/software/CBF/doc/) so making the minimum value be the undefined value made sense when dxtbx was laid down.  However, now that we have underload we should use it in our as our trusted minimum value if available, and further it would be good to use `undefined_value` when we create masks from Full CBFs.

2. In 1.8.4, [`overload`](http://www.bernstein-plus-sons.com/software/CBF/doc/cif_img_1.8.4.html#_array_intensities.overload) is  defined as "The saturation intensity level for this data array, i.e. the value above which correct intensities may not be recorded."  Both overload and underload include this phrase: "The valid pixel values are those less than _array_intensities.overload and greater than or equal to _array_intensities.underload."  To me these sentences are contradictory, so I got word from the author @yayahjb, who told me the section should read "The saturation intensity level for this data array, i.e. the value **_at or_** above which correct intensities may not be recorded. The valid pixel values are those less than _array_intensities.overload and greater than or equal to _array_intensities.underload"

What all this means is that, taking into account #536, [`imgCIF_H`](https://github.com/cctbx/dxtbx/blob/main/src/dxtbx/model/detector.py#L777-L778) needs to change in order to match the ImageCIF spec, from

```
            underload = find_undefined_value(cbf_handle)
            overload = cbf_handle.get_overload(0)
```
to
```
            trusted_min = find_underload_value(cbf_handle) # putative new function
            if trusted_min is None: trusted_min = find_undefined_value(cbf_handle) + 1
            trusted_max = cbf_handle.get_overload(0) - 1
```
This could break a lot of stuff, to say the least.

### PR DESCRIPTION
**Pull request by phyy-nx**
Fixes #559

Thanks @rjgildea for #587.  This is the other bit needed, reading the underload, and falling back to undefined if underload isn't present.

### CODE REVIEW

File: src/dxtbx/model/detector.py [hunk 1]
```diff
@@ -774,13 +775,26 @@ def imgCIF_H(cbf_handle, sensor):
...
+
+        try:
             # In imgCIF overload means the pixel is saturated and hence untrusted
             # https://www.iucr.org/__data/iucr/cifdic_html/2/cif_img.dic/Iarray_intensities.overload.html
-            overload = cbf_handle.get_overload(0)
-            trusted_range = (undefined_value + 1, (overload - 1) * dxtbx_overload_scale)
+            overload = (cbf_handle.get_overload(0) - 1) * dxtbx_overload_scale
         except Exception:
-            trusted_range = (0.0, 1.0e6)
+            overload = 1.0e6
```
Comment (rjgildea) [inline src/dxtbx/model/detector.py:792]: This makes me wonder is there a specific reason for this particualr fallback value?

**Review by rjgildea — APPROVED** [review]
Thanks for adding this, mostly looks good, just a few minor suggestions regarding variable names, and where the logic lies for checking the undefined value.

File: src/dxtbx/model/detector.py (hunk 1 above)
Comment (phyy-nx) [reply src/dxtbx/model/detector.py:790]: Good question!  Some digging shows it used to be 0, 0 which was not so useful, so @graeme-winter changed it in eb05fbc5fcb630f48ad724fd742f09f907f31cda.

**Comment by phyy-nx** [comment]
Good suggestions, thanks @rjgildea

File: src/dxtbx/model/detector.py (hunk 1 above)
Comment (graeme-winter) [reply src/dxtbx/model/detector.py:790]: No specific reason why 1e6: it is a generic big number

File: src/dxtbx/model/detector.py [hunk 2]
```diff
@@ -774,13 +775,21 @@ def imgCIF_H(cbf_handle, sensor):
         size = tuple(reversed(cbf_handle.get_image_size(0)))
 
         try:
-            undefined_value = find_undefined_value(cbf_handle)
+            underload = find_underload_value(cbf_handle)
+        except Exception:
+            try:
+                underload = find_undefined_value(cbf_handle) + 1
```
Comment (dagewa) [inline src/dxtbx/model/detector.py:781]: Came here looking from #595. Sorry I did not comment here earlier before this was merged, but I find this line totally mysterious. Why should the "undefined value" be one less than the "underload"? This sounds like a convention, so is there something that could be pointed to by a comment?

Also, I find the terms "underload" and "overload" to be confusing. It sounds like these are outside the trusted range, whereas what is actually meant here is `min_trusted_value` and `max_trusted_value`, right?

Comment (phyy-nx) [reply src/dxtbx/model/detector.py:781]: Agree on the first point.  I'll add some comments, over in #595.

The second one is harder, I'll also comment in #595.

**Pull request merged**